### DISCUSSION TITLE
3D non-cartesian sequence

### DISCUSSION THREAD

**Discussion by aTrotier** [JuliaHealth/KomaMRI.jl#391] [Q&A] [answered]
Hi,
I have a 3D UTE sequence that I want to simulate with Koma.
I am able to perform the simulation but the raw object only contains the kx/ky (not kz) trajectory even if I changed the obj_ui to phantom3D.

How can I solve this problem ?

**Comment by cncastillo** [comment]
In theory this should be possible by calling `signal_to_raw_data` with `; ndims=3`, but I haven't tested it yet. This is not easily doable in the UI, as it is hardcoded to `ndims=2`. I could put is as a optional/experimental argument in `KomaUI`.

@curtcorum was trying to fix the recon of 3D sequences in:

- #324 

But we were getting some weird errors, like:

```julia
julia> image = reconstruction(acq, reconParams);
ERROR: ArgumentError: Nodes x have dimension 2 != 3
```

I will try to take a look at this on the weekend!

**Comment by aTrotier** [comment]
Make sense, I think I should start using more and more the script for simulation now :)

In order to reconstruct a real 3D UTE acquisition from pulseq with MRIReco, I also need to edit some parameters in the raw that are wrongly encoded by Siemens (I guess some of them are deduced by the mdh). The trajectory dimension is one of it.

```
for p in eachindex(raw.profiles)
  raw.profiles[p].traj = Float32.(traj[:,:,p])
  raw.profiles[p].head.trajectory_dimensions=3
end
raw.params["trajectory"] = "custom"
raw.params["encodedSize"] = (128,128,128)
raw.params["fov"] = (256,256,256)
```

Right now I write the trajectory into the definition section of the .seq files from the seq.calculate_kspace().
Then I read the .seq with Koma in order edit the raw.profiles[i].traj.

The .seq size increase a lot. I think I can use Koma to generate the trajectory ?

**Reply by cncastillo** [reply, marked as answer]
Yes! you should be able to calculate it by using the function  `get_kspace`

```julia
kspace, kspace_adc = get_kspace(seq)
```
In most cases, the value of the $k$-space coordinates sampled at the ADC ponts is enough

```julia
_, kspace = get_kspace(seq)
```
This should match Pulseq's $k$-space definition for most sequences.

In the case of UTE sequences, the $k$-space could differ. As the calculation of the $k$-space depends on the calculation of the "RF center". In Koma, we define the RF center as the "center of mass" of the RF, in Pulseq it is the time where the maximum |B1| ocurrs (if I remember correctly, but I could be wrong). 

For half-sinc RFs, and VERSE half-sinc's this could affect the $k$-space calculations, as the proper RF center is the end of the pulse ($\boldsymbol{k}=\boldsymbol{0}$ happens at the end of the RF). Therefore, the $k_z$ component could have an offset. Having said that, I don't think this should affect the reconstruction, just add an small phase in the image.

**Reply by aTrotier** [reply]
Good to know I'll try, next week and compare the results.

When the pulseq branch is ready, I'll be happy to switch totally in Julia. I can do the 2 implementations and perform the acquisition if you need a real comparison

**Reply by aTrotier** [reply]
@cncastillo thanks it gives the same results and I dont have to write the traj into the .seq now :)

**Reply by sabbir-ahmed12** [reply]
Hi @aTrotier, did you implement 3D radial UTE sequence for your task? I've developed a phantom and wanted to try 3D radial UTE sequence for simulation. However, I don't have much knowledge about pulse sequence development. It'd be helpful if I could use an already available sequence to test the phantom.

**Reply by aTrotier** [reply]
Hi,

I have one implementation but did not make a lot of efforts to validate it.

Maybe you can take a look at this implementation :
- https://github.com/pdhe/Pulseq3DUTE

**Reply by sabbir-ahmed12** [reply]
Thanks for sharing!
